pkg/builder/kaniko/kaniko.go

```diff
@@ -0,0 +1,248 @@
+		return "", fmt.Errorf("error retrieving container logs: %w", err)
+	}
+
+	return string(logs), nil
+}
+
+func (k *Kaniko) cleanup(ctx context.Context, job *batchv1.Job) error {
+	err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).
+		Delete(ctx, job.Name, metav1.DeleteOptions{
+			PropagationPolicy: &[]metav1.DeletionPropagation{metav1.DeletePropagationBackground}[0],
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Job: %w", err)
+	}
+
+	// Delete the associated Pods
+	err = k.k8sClientset.CoreV1().Pods(k.k8sNamespace).
+		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("job-name=%s", job.Name),
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Pods: %w", err)
+	}
+
+	return nil
+}
+
+func GetDefaultCacheOptions(buildContext string) (*builder.CacheOptions, error) {
+	if buildContext == "" {
+		return nil, fmt.Errorf("build context cannot be empty")
+	}
+	hash := sha256.New()
+	_, err := hash.Write([]byte(buildContext))
+	if err != nil {
+		return nil, err
+	}
+	hashStr := hex.EncodeToString(hash.Sum(nil))
+
+	return &builder.CacheOptions{
+		Enabled: true,
+		Dir:     "",
+		Repo:    fmt.Sprintf("ttl.sh/%s:24h", hashStr),
+	}, nil
+}
+
+func prepareJob(b *builder.BuilderOptions) (*batchv1.Job, error) {
+	jobName, err := generateK8sName(kanikoJobNamePrefix)
+	if err != nil {
+		return nil, fmt.Errorf("error generating Job name: %w", err)
+	}
+
+	oneInt32 := int32(1)
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: jobName,
+		},
+		Spec: batchv1.JobSpec{
+			Parallelism: &oneInt32, // Set parallelism to 1 to ensure only one Pod
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  kanikoContainerName,
+							Image: kanikoImage, // debug has a shell
+							Args: []string{
+								`--context=` + b.BuildContext,
+								// TODO: see if we need it or not
+								// --git gitoptions    Branch to clone if build context is a git repository (default branch=,single-branch=false,recurse-submodules=false)
+
+								// TODO: we might need to add some options to get the auth token for the registry
+								"--destination=" + b.Destination,
+							},
+						},
+					},
+					RestartPolicy: "IfFailed",
+				},
+			},
+		},
+	}
+
+	// TODO: we need to add some configs to get the auth token for the cache repo
+	if b.Cache != nil && b.Cache.Enabled {
+		cacheArgs := []string{"--cache=true"}
+		if b.Cache.Dir != "" {
+			cacheArgs = append(cacheArgs, "--cache-dir="+b.Cache.Dir)
+		}
+		if b.Cache.Repo != "" {
+			cacheArgs = append(cacheArgs, "--cache-repo="+b.Cache.Repo)
+		}
+		job.Spec.Template.Spec.Containers[0].Args = append(job.Spec.Template.Spec.Containers[0].Args, cacheArgs...)
+	}
+
+	// Add extra args
+	job.Spec.Template.Spec.Containers[0].Args = append(job.Spec.Template.Spec.Containers[0].Args, b.Args...)
+
+	return job, nil
+
+}
+
+func generateK8sName(name string) (string, error) {
```

Comment:
feels like generating pod / resource names should maybe live on its own in a package? eg: `names.NewRandom()`

Reply:
Good idea we can do that and use it across the whole repo